src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+
+        private PipeCompletion _writerCompletion;
+        private PipeCompletion _readerCompletion;
+
+        private readonly BufferSegment[] _bufferSegmentPool;
+
+        // The read head which is the extent of the IPipelineReader's consumed bytes
+        private BufferSegment _readHead;
+        private int _readHeadIndex;
+
+        // The commit head which is the extent of the bytes available to the IPipelineReader to consume
+        private BufferSegment _commitHead;
+        private int _commitHeadIndex;
+
+        // The write head which is the extent of the IPipelineWriter's written bytes
+        private BufferSegment _writingHead;
+
+        private PipeReaderState _readingState;
+
+        private bool _disposed;
+
+        private DefaultPipeReader _reader;
+        private DefaultPipeWriter _writer;
+
+        internal long Length => _length;
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> using <see cref="PipeOptions.Default"/> as options.
+        /// </summary>
+        public Pipe() : this(PipeOptions.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="Pipe"/> with the specified <see cref="PipeOptions"/>.
+        /// </summary>
+        public Pipe(PipeOptions options)
+        {
+            if (options == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.options);
+            }
+
+            _bufferSegmentPool = new BufferSegment[SegmentPoolSize];
+
+            _readingState = default;
+            _readerCompletion = default;
+            _writerCompletion = default;
+
+            _pool = options.Pool;
+            _minimumSegmentSize = options.MinimumSegmentSize;
+            _maximumSizeHigh = options.PauseWriterThreshold;
+            _maximumSizeLow = options.ResumeWriterThreshold;
+            _readerScheduler = options.ReaderScheduler ?? PipeScheduler.Inline;
+            _writerScheduler = options.WriterScheduler ?? PipeScheduler.Inline;
+            _readerAwaitable = new PipeAwaitable(completed: false);
+            _writerAwaitable = new PipeAwaitable(completed: true);
+            _reader = new DefaultPipeReader(this);
+            _writer = new DefaultPipeWriter(this);
+        }
+
+        private void ResetState()
+        {
+            _readerCompletion.Reset();
+            _writerCompletion.Reset();
+            _commitHeadIndex = 0;
+            _currentWriteLength = 0;
+            _length = 0;
+        }
+
+        /// <summary>
+        /// Allocates memory from the pipeline to write into.
+        /// </summary>
+        /// <param name="minimumSize">The minimum size buffer to allocate</param>
+        internal Memory<byte> GetMemory(int minimumSize)
+        {
+            if (_writerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed();
+            }
+
+            if (minimumSize < 0)
```

Comment:
Is zero ok? Why would we allow that? In fact, whenever I write code using our IOutput APIs I need to have a special branch for empty buffer.

Reply:
Zero is whatever is availible at the moment but not empty

Reply:
1 is also whatever is available at the moment but not zero, correct? If yes, then default being one and zero throwing would be more transparent API.

Reply:
No, one is "I need to write one byte", zero is for "whatever is availible"